R/excitation.R

```diff
@@ -0,0 +1,147 @@
+#' Excitation
+#'
+#' @description Excitation lets you add visual citations as tooltips to your
+#'   \pkg{xaringan} slides. Simply include a `.bib` file in your R Markdown
+#'   YAML header, as usual, and cite away!
+#'
+#' @section Usage: To use excitation for visual citations, add the following
+#'   code chunk to your slides' R Markdown file.
+#'
+#'   ````markdown
+#'   ```{r xaringan-excitation, echo=FALSE}
+#'   xaringanExtra::use_excitation()
+#'   ```
+#'   ````
+#'
+#' @examples
+#' use_excitation()
+#'
+#' @name excitation
+NULL
+
+#' @describeIn excitation Adds visual citations to your xaringan slides.
+#' @export
+use_excitation <- function() {
+  assign('.excitation_keys', vector('character'), envir = globalenv())
+  htmltools::tagList(
+    html_dependency_popper(),
+    html_dependency_tippy(),
+    html_dependency_excitation()
+  )
+}
+
+#' @describeIn excitation Adds visual citations to your xaringan slides.
+#' @export
+excite <- function(key) {
+  bibs <- rmarkdown::metadata$bibliography
+  self_contained <- FALSE
+
+  if (is.null(bibs)) return(key)
+
+  if (length(bibs) > 1) {
+    tmp_bib <- tempfile(fileext = '.bib')
+    for (i in seq_len(bibs)) {
+      write(readLines(bibs[[i]]), tmp_bib, append = TRUE)
+    }
+    bibs <- tmp_bib
+  }
```

Comment:
This part should be done only once and stored in a package env, so maybe the internal env would be `.excitation` with `.excitation$keys` and `.excitation$bib_file`.

Reply:
Actually, I think what you'd want to do, in order to make sure we don't ever leave any files lying around, is to read the bibfiles into a character string on the first run. Then for each `excite()` call, write the bib file to a temp file and make sure you clean it up with `on.exit(unlink(tmp_bib))` right after writing the file.

I'm also realizing you want to add some logic around the caching mechanism in case people use different bibliographies for each `excite()` call.

Reply:
Hmm, do you think users would supply a different bib file when making each individual citation (i.e. calling `excite()`)? I am unsure of the use case.

Maybe the bib stuff should be handled in `use_excitation()`, where it has the argument `bib = NULL` and defaults to using the YAML `bibliography` entry.

Following from your second comment, we could store the bib(s) in a string in the internal environment and then write the bib file and clean it up in `excite()` for each call. This removes the need for any caching logic.

Reply:
Yeah doing it through `use_excitation()` which inherits from the yaml makes a lot of sense. But I think it'll be worth the effort to add an arg here. On the one hand it becomes easier to test `excite()` since you can set test it without relying on side effects. On the other hand, in my experience functions that are easier to test are also more flexible and easier to use in more places.

I think we can compromise, though, and say that only the default is cached. If `bib` is provided by the user it would be a path that you can just pass along to `drop_name()`. I think that'd be a good balance between flexibility and not writing too much code 😉 

(Also I'm not attached to `bib` as the parameter name; `bibliography` would work too. Whatever makes the most sense.)

Reply:
Sorry, just to clarify, you think we should have a `bib` arg for `excite()` only, or for both functions?

Reply:
Both. `use_excitation()` could take a vector of bib files and use caching to minimize file access. `excited()` on the other hand would either pass `bib` directly to `drop_name()` if provided by the user or would use the default bib file(s) if not specified.